File: example/sarsa_offense/high_level_sarsa_agent.cpp
#include <iostream>
#include <vector>
#include <HFO.hpp>
#include <cstdlib>
#include <thread>
#include "SarsaAgent.h"
#include "CMAC.h"
#include <unistd.h>

// Before running this program, first Start HFO server:
// $./bin/HFO --offense-agents numAgents

void printUsage() {
  std::cout<<"Usage: ./high_level_sarsa_agent [Options]"<<std::endl;
  std::cout<<"Options:"<<std::endl;
  std::cout<<"  --numAgents <int>        Number of SARSA agents"<<std::endl;
  std::cout<<"                           Default: 1"<<std::endl;
  std::cout<<"  --numEpisodes <int>      Number of episodes to run"<<std::endl;
  std::cout<<"                           Default: 10"<<std::endl;
  std::cout<<"  --basePort <int>         SARSA agent base port"<<std::endl;
  std::cout<<"                           Default: 6000"<<std::endl;
  std::cout<<"  --learnRate <float>      Learning rate of SARSA agents"<<std::endl;
  std::cout<<"                           Range: [0.0, 1.0]"<<std::endl;
  std::cout<<"                           Default: 0.1"<<std::endl;
  std::cout<<"  --suffix <int>           Suffix for weights files"<<std::endl;
  std::cout<<"                           Default: 0"<<std::endl;
  std::cout<<"  --noOpponent             Sets opponent present flag to false"<<std::endl;
  std::cout<<"  --help                   Displays this help and exit"<<std::endl;
}

// Returns the reward for SARSA based on current state
double getReward(int status) {
  double reward;
  if (status==hfo::GOAL) reward = 1;
  else if (status==hfo::CAPTURED_BY_DEFENSE) reward = -1;
  else if (status==hfo::OUT_OF_BOUNDS) reward = -1;
  else reward = 0;
  return reward;
}

// Fill state with only the required features from state_vec
void purgeFeatures(double *state, const std::vector<float>& state_vec,
                   int numTMates, bool oppPres) {

  int stateIndex = 0;

  // If no opponents ignore features Distance to Opponent
  // and Distance from Teammate i to Opponent are absent
  int tmpIndex = 9 + 3 * numTMates;

  for(int i = 0; i < state_vec.size(); i++) {

    // Ignore first six features and teammate proximity to opponent(when opponent is absent)and opponent features
    if(i < 6||(!oppPres && ((i>9+numTMates && i<=9+2*numTMates)||i==9))||i>9+6*numTMates) continue;

    // Ignore Angle and Uniform Number of Teammates
    int temp =  i-tmpIndex;
    if(temp > 0 && (temp % 3 == 2 || temp % 3 == 0)) continue;
    if (i > 9+6*numTMates) continue;
    state[stateIndex] = state_vec[i];
    stateIndex++;
  }
  //std::cout<<stateIndex<<"yo";
}

// Convert int to hfo::Action
hfo::action_t toAction(int action, const std::vector<float>& state_vec) {
  hfo::action_t a;
  switch(action) {
    case 0: a = hfo::SHOOT;
            break;
    case 1: a = hfo::DRIBBLE;
            break;
    default:int size = state_vec.size();
            a = hfo::PASS;/*,
                 state_vec[(size - 1) - (action - 2) * 3],
                 0.0};*/
  }
  return a;
}

void offenseAgent(int port, int numTMates, int numEpi, double learnR,
                  int suffix, bool oppPres, double eps) {

  // Number of features
  int numF = oppPres ? (4 + 4 * numTMates) : (3 + 3 * numTMates);
  // Number of actions
  int numA = 2 + numTMates;

  double discFac = 1;

  // Tile coding parameter
  double resolution = 0.1;

  double range[numF];
  double min[numF];
  double res[numF];
  for(int i = 0; i < numF; i++) {
      min[i] = -1;
      range[i] = 2;
      res[i] = resolution;
  }

  // Weights file
  char *wtFile;
  std::string s = "weights_" + std::to_string(port) +
                  "_" + std::to_string(numTMates + 1) +
                  "_" + std::to_string(suffix);
  wtFile = &s[0u];
  double lambda = 0;
  CMAC *fa = new CMAC(numF, numA, range, min, res);
  SarsaAgent *sa = new SarsaAgent(numF, numA, learnR, eps, lambda, fa, wtFile, wtFile);

  hfo::HFOEnvironment hfo;
  hfo::status_t status;
  hfo::action_t a;
  double state[numF];
  int action = -1;
  double reward;
  hfo.connectToServer(hfo::HIGH_LEVEL_FEATURE_SET,"../../bin/teams/base/config/formations-dt",6000,"localhost","base_left",false,"");
  for (int episode=0; episode < numEpi; episode++) {
    int count = 0;
    status = hfo::IN_GAME;
    action = -1;
    while (status == hfo::IN_GAME) {
      const std::vector<float>& state_vec = hfo.getState();
      // If has ball
      if(state_vec[5] == 1) {
        if(action != -1) {
          reward = getReward(status);
          sa->update(state, action, reward, discFac);
        }

        // Fill up state array
        purgeFeatures(state, state_vec, numTMates, oppPres);

	// Get raw action
        action = sa->selectAction(state);

        // Get hfo::Action
        a = toAction(action, state_vec);

      } else {
            a = hfo::MOVE;
      }
      if (a== hfo::PASS) {
           hfo.act(a,state_vec[(9+6*numTMates) - (action-2)*3]);
           //std::cout<<(9+6*numTMates) - (action-2)*3;
        } else {
           hfo.act(a);
        }
      status = hfo.step();
    }
    // End of episode
    if(action != -1) {
      reward = getReward(status);
      sa->update(state, action, reward, discFac);
      sa->endEpisode();
    }
  }

  delete sa;
  delete fa;
}

int main(int argc, char **argv) {

  int numAgents = 1;
  int numEpisodes = 10;
  int basePort = 6000;
  double learnR = 0.1;
  int suffix = 0;
  bool opponentPresent = true;
  double eps = 0.01;
  for(int i = 1; i < argc; i++) {
    std::string param = std::string(argv[i]);
    if(param == "--numAgents") {
      numAgents = atoi(argv[++i]);
    }else if(param == "--numEpisodes") {
      numEpisodes = atoi(argv[++i]);
    }else if(param == "--basePort") {
      basePort = atoi(argv[++i]);
    }else if(param == "--learnRate") {
      learnR = atof(argv[++i]);
      if(learnR < 0 || learnR > 1) {
        printUsage();
        return 0;
      }
    }else if(param == "--suffix") {
      suffix = atoi(argv[++i]);
    }else if(param == "--noOpponent") {
      opponentPresent = false;
    }else if(param=="--eps"){
        eps=atoi(argv[++i]);
    }else {
      printUsage();
      return 0;
    }
  }

  int numTeammates = numAgents - 1;

  std::thread agentThreads[numAgents];
  for (int agent = 0; agent < numAgents; agent++) {
    agentThreads[agent] = std::thread(offenseAgent, basePort + agent,
                                      numTeammates, numEpisodes, learnR,
                                      suffix, opponentPresent, eps);
    usleep(500000L);
  }
  for (int agent = 0; agent < numAgents; agent++) {
    agentThreads[agent].join();
  }
  return 0;
}

File: example/sarsa_offense/high_level_sarsa_agent.py
#!/usr/bin/env python3
# encoding: utf-8

from hfo import *
import argparse
import numpy as np
import sys, os
sys.path.append(os.path.join(os.path.dirname(__file__), '..', 'sarsa_libraries','python_wrapper'))
from py_wrapper import *

def getReward(s):
  reward=0
  #--------------------------- 
  if s==GOAL:
    reward=1
  #--------------------------- 
  elif s==CAPTURED_BY_DEFENSE:
    reward=-1
  #--------------------------- 
  elif s==OUT_OF_BOUNDS:
    reward=-1
  #--------------------------- 
  #Cause Unknown Do Nothing
  elif s==OUT_OF_TIME:
    reward=0
  #--------------------------- 
  elif s==IN_GAME:
    reward=0
  #--------------------------- 
  elif s==SERVER_DOWN:  
    reward=0
  #--------------------------- 
  else:
    print("Error: Unknown GameState", s)
  return reward

def purge_features(state):
  st=np.empty(NF,dtype=np.float64)
  stateIndex=0
  tmpIndex= 9 + 3*NOT
  for i in range(len(state)):
    # Ignore first six features and teammate proximity to opponent(when opponent is absent)and opponent features
    if(i < 6 or i>9+6*NOT or (NOO==0 and ((i>9+NOT and i<=9+2*NOT) or i==9)) ): 
      continue;
    #Ignore Angle and Uniform Number of Teammates
    temp =  i-tmpIndex;
    if(temp > 0 and (temp % 3 == 2 or temp % 3 == 0)):
       continue;
    if (i > 9+6*NOT): 
      continue;
    st[stateIndex] = state[i];
    stateIndex+=1;
  return st


if __name__ == '__main__':
  parser = argparse.ArgumentParser()
  parser.add_argument('--port', type=int, default=6000)
  parser.add_argument('--numTeammates', type=int, default=0)
  parser.add_argument('--numOpponents', type=int, default=1)
  parser.add_argument('--numEpisodes', type=int, default=1)
  parser.add_argument('--learnRate', type=float, default=0.1)
  parser.add_argument('--suffix', type=int, default=0)
  args=parser.parse_args()

  # Create the HFO Environment
  hfo = HFOEnvironment()
  #now connect to the server
  hfo.connectToServer(HIGH_LEVEL_FEATURE_SET,'bin/teams/base/config/formations-dt',args.port,'localhost','base_left',False)
  global NF,NA,NOT,NOO
  NOO=args.numOpponents
  if args.numOpponents >0: 
    NF=4+4*args.numTeammates
  else:
    NF=3+3*args.numTeammates
  NOT=args.numTeammates
  NA=NOT+2 #PASS to each teammate, SHOOT, DRIBBLE
  learnR=args.learnRate
  #CMAC parameters
  resolution=0.1
  Range=[2]*NF
  Min=[-1]*NF
  Res=[resolution]*NF
  #Sarsa Agent Parameters
  wt_filename="weights_"+str(NOT+1)+"v"+str(NOO)+'_'+str(args.suffix)
  discFac=1
  Lambda=0
  eps=0.01
  #initialize the function approximator and the sarsa agent
  FA=CMAC(NF, NA, Range, Min, Res)
  SA=SarsaAgent(NF, NA, learnR, eps, Lambda, FA, wt_filename, wt_filename)
 
  #episode rollouts 
  st = np.empty(NF,dtype=np.float64)
  action = -1
  reward = 0
  for episode in range(1,args.numEpisodes+1):
    count=0
    status=IN_GAME
    action=-1

    while status==IN_GAME:
      count=count+1
      # Grab the state features from the environment
      state = hfo.getState()
      if int(state[5])==1:
        if action != -1:
          #print(st)
          reward=getReward(status)
          #fb.SA.update(state,action,reward,discFac)
          SA.update(st,action,reward,discFac)
        st=purge_features(state)
        #take an action
        #action = fb.SA.selectAction(state)
        action = SA.selectAction(st)
        #print("Action:", action)
        if action == 0:
          hfo.act(SHOOT) 
        elif action == 1:
          hfo.act(DRIBBLE)
        else:
          hfo.act(PASS,state[(9+6*NOT)-(action-2)*3]) 
      else:
        hfo.act(MOVE)
        
      status = hfo.step()
    #--------------- end of while loop ------------------------------------------------------
    
    ############# EPISODE ENDS ###################################################################################
    # Check the outcome of the episode
    if action != -1:
      reward=getReward(status)
      SA.update(st, action, reward, discFac)
      SA.endEpisode()
    ############################################################################################################
    # Quit if the server goes down
    if status == SERVER_DOWN:
      hfo.act(QUIT)
      break

   
